### PR TITLE
Tasks refactor

### PR DESCRIPTION
**Pull request by swarajpure**
- [x] New task statuses should be visible in sidebar
- [x] New task statuses should be selectable in a dropdown for a task, such that task status can be changed by clicking the dropdown
- [x] Message in case of no tasks being found
- [x] Refactor logic for filtering tasks according to task type
- [x] Refactor logic after API call to update a task
- [x] Add loader and response after updating a task
- [x] Add ALL tasks in the sidebar such that we support and show the tasks with older task statuses

### CODE REVIEW

**Comment by swarajpure** [comment]
`All tasks` is shown by default. Notice that it contains current task status as well:
image

`Assigned tasks` view. Notice that the task slider isn't shown here:
image

`In Progress` tasks view. Notice that we're showing task slider here:
image

`No task found` message:
image

File: app/components/side-bar.hbs [hunk 1]
```diff
@@ -3,7 +3,7 @@
   <FaIcon @icon="caret-down" @fixedWidth={{true}} />  </button>
   <div class="sidebar__dropdown-container {{if @showDropDown "show" "close"}}">
     {{#each @taskStatusList as |status|}}
-      <a href="#" {{on "click" (fn @toggleTasks status)}}>{{status}}</a>
+      <a href="#" {{on "click" (fn @changeUserSelectedTask status)}}>{{status.label}}</a>
```
Comment (ankushdharkar) [inline app/components/side-bar.hbs:6]: This should be a `<button>`.
Refer: https://www.webaxe.org/proper-use-buttons-links/

File: app/components/task/holder.hbs [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
-      <input
-        id="input"
-        type="range"
-        name="percentCompleted"
-        value="{{@task.percentCompleted}}"
-        min="0" max="100" step="25"
-        {{on "change" this.onPercentageChange}}
-        {{on "change" (fn @onTaskUpdate @task.id)}}
-      />
+  {{#if (eq @task.status 'IN_PROGRESS')}}
```
Comment (ankushdharkar) [inline app/components/task/holder.hbs:3]: Can we use a known enum here please?

File: app/components/task/holder.hbs [hunk 3]
```diff
@@ -1,16 +1,42 @@
...
+    <div class="progress-bar-container">
```
Comment (ankushdharkar) [inline app/components/task/holder.hbs:4]: Are we breaking BEM?

File: app/components/task/holder.js [hunk 4]
```diff
@@ -17,5 +15,5 @@ export default class TasksHolderComponent extends Component {
     this.args.onTaskChange('status', value);
   }
 
-  availableStatusOptions = [ACTIVE, BLOCKED, COMPLETED, PENDING];
+  availableTaskStatuses = TASK_STATUSES;
```
Comment (ankushdharkar) [inline app/components/task/holder.js:18]: Should this be moved above actions for readability?

File: app/constants/tasks.js [hunk 5]
```diff
@@ -1,8 +1,64 @@
-const STATUS = {
-  ACTIVE: 'active',
-  BLOCKED: 'blocked',
-  COMPLETED: 'completed',
-  PENDING: 'pending',
-};
+const TASK_STATUSES = [
```
Comment (ankushdharkar) [inline app/constants/tasks.js:1]: I wonder if this should be a map, instead of an array.

If I want to look up if there's a key called `TASK_STATUSES.<keyname>`, we have no easy way of doing it anymore. What do we benefit from making this an array here?

File: app/controllers/tasks.js [hunk 6]
```diff
@@ -1,67 +1,61 @@
...
-  filterTasksByStatus(allTasks, status) {
-    return allTasks.filter((task) => task.status === status);
-  }
-
-  seperateTasksByStatus() {
-    this.taskStatusList.forEach((status) => {
-      this.tasksByStatus[status] = this.filterTasksByStatus(
-        this.allTasks,
-        status
+  calculateTasksToShowBySelectedStatus() {
```
Comment (ankushdharkar) [inline app/controllers/tasks.js:24]: Would this be better as a `get`?

File: app/controllers/tasks.js [hunk 7]
```diff
@@ -1,67 +1,61 @@
...
+    if (this.userSelectedTask.key === 'all') {
```
Comment (ankushdharkar) [inline app/controllers/tasks.js:25]: Maybe use constant here?

File: app/controllers/tasks.js [hunk 8]
```diff
@@ -1,67 +1,61 @@
...
   @action async handleUpdateTask(taskId) {
     const taskData = this.taskFields;
-    if (taskData.status === COMPLETED) {
-      taskData.percentCompleted = 100;
-    }
     const cleanBody = this.cleanReqBody(taskData);
     if (taskData.status || taskData.percentCompleted) {
       try {
+        this.isLoading = true;
         const response = await fetch(`${API_BASE_URL}/tasks/self/${taskId}`, {
```
Comment (ankushdharkar) [inline app/controllers/tasks.js:59]: This should be moved in ember-data.

File: app/controllers/tasks.js [hunk 9]
```diff
@@ -1,67 +1,61 @@
...
 
   @action async handleUpdateTask(taskId) {
     const taskData = this.taskFields;
-    if (taskData.status === COMPLETED) {
-      taskData.percentCompleted = 100;
-    }
     const cleanBody = this.cleanReqBody(taskData);
     if (taskData.status || taskData.percentCompleted) {
       try {
+        this.isLoading = true;
```
Comment (ankushdharkar) [inline app/controllers/tasks.js:58]: Do we need this in try? Try should contain the minimal section of code that it can.

File: app/styles/tasks.css [hunk 10]
```diff
@@ -189,7 +189,17 @@ input[type=range]:focus::-ms-fill-upper {
...
+  margin-bottom: 15px;
+}
+
+.no-tasks-found {
+  margin-top: 20px;
+}
 
+.loader-container {
+  margin-top: 10px;
+}
```
Comment (ankushdharkar) [inline app/styles/tasks.css:202]: BEM?

File: tests/integration/components/tasks-test.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ module('Integration | Component | tasks', function (hooks) {
```
Comment (ankushdharkar) [inline tests/integration/components/tasks-test.js:12]: Do we need more tests to cover the new feature(s)?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: app/components/side-bar.hbs (hunk 1 above)
Comment (swarajpure) [reply app/components/side-bar.hbs:6]: Oh yes, my bad! Missed it

File: app/components/task/holder.hbs (hunk 2 above)
Comment (swarajpure) [reply app/components/task/holder.hbs:3]: Didn't get this point, couldn't find anything useful on Googling for `enum in ember` or `enum in handlebars`
Can you please provide some resource for this one?

File: app/constants/tasks.js (hunk 5 above)
Comment (swarajpure) [reply app/constants/tasks.js:1]: we're using this array to show the sidebar options, if we use map, the order would be messed up, right?

File: app/controllers/tasks.js (hunk 8 above)
Comment (swarajpure) [reply app/controllers/tasks.js:59]: i agree but ember data isn't in place yet

File: tests/integration/components/tasks-test.js (hunk 11 above)
Comment (swarajpure) [reply tests/integration/components/tasks-test.js:12]: yes, we definitely need tests, there are no tests as of now, since this feature had a tight deadline

**Comment by swarajpure** [comment]
Normal view:
image

In case of hovering over any task status:
image

File: app/components/task/holder.hbs [hunk 12]
```diff
@@ -1,16 +1,44 @@
-<div class="task-card">
+<div class="task-card" tabindex="0">
```
Comment (swarajpure) [inline app/components/task/holder.hbs:1]: added `tabindex` so each task card can be highlighted when traversing via keyboard

**Review by Pavangbhat — COMMENTED** [review]
Hi @swarajpure 
Are we using **/tasks/self** endpoint  to get all the tasks, if yes we cannot make use of it since it does not return all the tasks 
Checkout this 👇 
https://github.com/Real-Dev-Squad/website-api-contracts/tree/main/tasks#get-tasksself

we can make use of **/tasks/username** 👇 
https://github.com/Real-Dev-Squad/website-api-contracts/tree/main/tasks#get-tasksusername

**RESOLVED ⬆️ **

File: app/constants/tasks.js (hunk 5 above)
Comment (ankushdharkar) [reply app/constants/tasks.js]: @swarajpure We can take the enum list, and make it into an array using [Object.keys](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys)

File: tests/integration/components/tasks-test.js (hunk 11 above)
Comment (ankushdharkar) [reply tests/integration/components/tasks-test.js:12]: @swarajpure Then let's please have a ticket that tracks the follow-up of tests with timeline/ETA

File: app/controllers/tasks.js (hunk 6 above)
Comment (ankushdharkar) [reply app/controllers/tasks.js]: What I mean is that here, `calculateTasksToShowBySelectedStatus` is being called imperatively. From what I see in its implementation, it is used to update the details of `tasksToShow`, which in turns depends on the `userSelectedTask.key`

We should be only setting the `userSelectedTask.key`, and have `tasksToShow` be 'derived' from the dependency(s). So, we derive the data, instead of having to remember to call the `calculateTasksToShowBySelectedStatus`.

Read more on[ tracked properties](https://guides.emberjs.com/release/upgrading/current-edition/tracked-properties/) and let me know if you still have any questions.

File: app/constants/tasks.js (hunk 5 above)
Comment (swarajpure) [reply app/constants/tasks.js]: My concern with using an object is, will we be able to ensure the order of the keys? If the order is messed up, the list will obtained from `Object.keys` will also have different order and the sidebar won't have predictable order. Someone might say "I remember `In Progress` was the third option last week, why and how is it changed now?" @ankushdharkar

File: app/controllers/tasks.js (hunk 6 above)
Comment (swarajpure) [reply app/controllers/tasks.js]: We've added functionality to view all tasks (no filtering based on task status)
Here's the function definition:
```JS
  filterTasksByStatus() {
    if (this.userSelectedTask.key === this.allTasksObject.key) {
      this.tasksToShow = this.allTasks;
    } else {
      this.tasksToShow = this.allTasks.filter(
        (task) => task.status === this.userSelectedTask.key
      );
    }
  }
```

File: app/controllers/tasks.js (hunk 8 above)
Comment (swarajpure) [reply app/controllers/tasks.js]: #200

File: app/constants/tasks.js (hunk 5 above)
Comment (ankushdharkar) [reply app/constants/tasks.js]: @swarajpure Good point. So we can do something like this:
```js
const DISPLAY_LIST = [ IN_PROGRESS, COMPLETED, UNASSIGNED ];
```

Why will this order change?

File: app/components/task/holder.hbs (hunk 2 above)
Comment (ankushdharkar) [reply app/components/task/holder.hbs:3]:
```js
const STATUSES = {
  IN_PROGRESS: 'IN_PROGRESS',
}

```

File: config/environment.js [hunk 13]
```diff
@@ -23,7 +23,7 @@ module.exports = function (environment) {
     },
   };
 
-  ENV.BASE_API_URL = 'https://api.realdevsquad.com';
+  ENV.BASE_API_URL = 'http://localhost:3000';
```
Comment (swarajpure) [inline config/environment.js:26]: Will revert the change
